core-api/src/main/java/com/optimizely/ab/notification/NotificationCenter.java

```diff
@@ -73,104 +72,63 @@ public Class getNotificationTypeClass() {
         }
     }
 
-    ;
-
-
-    // the notification id is incremented and is assigned as the callback id, it can then be used to remove the notification.
-    private AtomicInteger notificationListenerID = new AtomicInteger();
+    private final AtomicInteger counter = new AtomicInteger();
+    private final ConcurrentHashMap<Class, NotificationManager> notifierMap = new ConcurrentHashMap<>();
 
-    final private static Logger logger = LoggerFactory.getLogger(NotificationCenter.class);
-
-    // notification holder holds the id as well as the notification.
-    private static class NotificationHolder {
-        int notificationId;
-        NotificationListener notificationListener;
-        DecisionNotificationListener decisionNotificationListener;
-
-        NotificationHolder(int id, NotificationListener notificationListener) {
-            notificationId = id;
-            this.notificationListener = notificationListener;
-        }
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public <T> NotificationManager<T> getNotificationManager(Class<? extends Notification> clazz) {
+        NotificationManager<T> newManager = new NotificationManager<>(counter);
+        NotificationManager<T> manager = (NotificationManager<T>) notifierMap.putIfAbsent(clazz, newManager);
```

Comment:
Use `computeIfAbsent()` to avoid creating throw-away instance when manager exists? Alternatively, the set of notification types is (all likely will remain) trivially small, so could just pre-populate the map.